geth/transactions/queue/queue.go

```diff
@@ -119,49 +119,21 @@ func (q *TxQueue) evictionLoop() {
 	}
 }
 
-// enqueueLoop process incoming enqueue requests
-func (q *TxQueue) enqueueLoop() {
-	defer HaltOnPanic()
-
-	// enqueue incoming transactions
-	for {
-		select {
-		case queuedTx := <-q.incomingPool:
-			log.Info("transaction enqueued requested", "tx", queuedTx.ID)
-			err := q.Enqueue(queuedTx)
-			log.Warn("transaction enqueued error", "tx", err)
-			log.Info("transaction enqueued", "tx", queuedTx.ID)
-		case <-q.stopped:
-			log.Info("transaction queue's enqueue loop stopped")
-			q.stoppedGroup.Done()
-			return
-		}
-	}
-}
-
 // Reset is to be used in tests only, as it simply creates new transaction map, w/o any cleanup of the previous one
 func (q *TxQueue) Reset() {
 	q.mu.Lock()
 	defer q.mu.Unlock()
 
 	q.transactions = make(map[common.QueuedTxID]*common.QueuedTx)
 	q.evictableIDs = make(chan common.QueuedTxID, DefaultTxQueueCap)
-}
-
-// EnqueueAsync enqueues incoming transaction in async manner, returns as soon as possible
-func (q *TxQueue) EnqueueAsync(tx *common.QueuedTx) error {
-	q.incomingPool <- tx
-
-	return nil
+	q.inprogress = make(map[common.QueuedTxID]empty)
 }
 
 // Enqueue enqueues incoming transaction
 func (q *TxQueue) Enqueue(tx *common.QueuedTx) error {
 	log.Info(fmt.Sprintf("enqueue transaction: %s", tx.ID))
-
-	if q.txEnqueueHandler == nil { //discard, until handler is provided
-		log.Info("there is no txEnqueueHandler")
-		return nil
+	if (tx.Hash != gethcommon.Hash{}) {
```

Comment:
What if a transaction errored? Before it would not have `Hash` set but `Error`.

It looks like this method is not thread-safe:
1. `tx.Hash` access is not guarded,
2. It's possible to enqueue already enqueued transaction (it looks like this was an issue before as well).

Reply:
i thought that it makes sense to enqueu transaction again, even if an error was not transient.
but to be fair, is it even required? i dont think that i've seen a code path that can actually try to enqueue same transaction again

Reply:
just for the reference, in this commit https://github.com/status-im/status-go/pull/537/commits/d75f6f02ba701da766f2472964e0683409abbc8b (PR https://github.com/status-im/status-go/pull/537), i removed this condition completely and using a channel to return error and a hash. 

but i am relying just on the tests, and i might be wrong if there is some history behind it

Reply:
Have you thought about some particular use cases? 

In our case, if a transaction reports an error, and it's not transient, then the transaction is removed and a signal is sent. If the error is transient, the transaction stays in the queue and a signal is sent. I don't recall any use cases that would allow or require putting the same transaction again to the queue, thus, I would throw an error in case of such an attempt.

Reply:
i will add error comparison for now, and will get back to this topic in https://github.com/status-im/status-go/pull/537